src/ring_theory/witt_vector/domain.lean

```diff
@@ -0,0 +1,200 @@
+begin
+  have hex : ∃ k : ℕ, x.coeff k ≠ 0,
+  { by_contradiction hall,
+    push_neg at hall,
+    apply hx,
+    ext i,
+    simp only [hall, zero_coeff] },
+  let n := nat.find hex,
+  use [n, x.shift n],
+  refine ⟨nat.find_spec hex, eq_iterate_verschiebung (λ i hi, not_not.mp _)⟩,
+  exact nat.find_min hex hi,
+end
+
+/-!
+## Iteration lemmas
+
+This construction involves iterated applications of `verschiebung` and `frobenius`.
+Here we prove some useful lemmas about these.
+Auxiliary lemmas are used to simplify double inductions.
+-/
+
+lemma iterate_verschiebung_coeff (x : 𝕎 R) (n k : ℕ) :
+  (verschiebung^[n] x).coeff (k + n) = x.coeff k :=
+begin
+  induction n with k ih,
+  { simp },
+  { rw [iterate_succ_apply', nat.add_succ, verschiebung_coeff_succ],
+    exact ih }
+end
+
+lemma iterate_verschiebung_mul_left (x y : 𝕎 R) (i : ℕ) :
+  (verschiebung^[i] x) * y = (verschiebung^[i] (x * (frobenius^[i] y))) :=
+begin
+  induction i with i ih generalizing y,
+  { simp },
+  { rw [iterate_succ_apply', ← verschiebung_mul_frobenius, ih, iterate_succ_apply'], refl }
+end
+
+/-!
+From this point on, we assume `R` is of characteristic `p`.
+-/
+
+section char_p
+
+variable [char_p R p]
+
+lemma iterate_verschiebung_mul (x y : 𝕎 R) (i j : ℕ) :
+  (verschiebung^[i] x) * (verschiebung^[j] y) =
+    (verschiebung^[i + j] ((frobenius^[j] x) * (frobenius^[i] y))) :=
+begin
+  calc
+  _ = (verschiebung^[i] (x * (frobenius^[i] ((verschiebung^[j] y))))) : _
+... = (verschiebung^[i] (x * (verschiebung^[j] ((frobenius^[i] y))))) : _
+... = (verschiebung^[i] ((verschiebung^[j] ((frobenius^[i] y)) * x))) : _
+... = (verschiebung^[i] ((verschiebung^[j] ((frobenius^[i] y) * (frobenius^[j] x))))) : _
+... = (verschiebung^[i + j] ((frobenius^[i] y) * (frobenius^[j] x))) : _
+... = _ : _,
+  { apply iterate_verschiebung_mul_left },
+  { rw verschiebung_frobenius_comm.iterate_iterate; apply_instance },
+  { rw mul_comm },
+  { rw iterate_verschiebung_mul_left },
+  { rw iterate_add_apply },
+  { rw mul_comm }
+end
+
+lemma iterate_frobenius_coeff (x : 𝕎 R) (i k : ℕ) :
+  ((frobenius^[i] x)).coeff k = (x.coeff k)^(p^i) :=
+begin
+  induction i with i ih,
+  { simp },
+  { rw [iterate_succ_apply', coeff_frobenius_char_p, ih],
+    ring_exp }
+end
+
+/-- This is a slightly specialized form of [Hazewinkel, *Witt Vectors*][Haze09] 6.2 equation 5. -/
+lemma iterate_verschiebung_mul_coeff (x y : 𝕎 R) (i j : ℕ) :
+  ((verschiebung^[i] x) * (verschiebung^[j] y)).coeff (i + j) =
+    (x.coeff 0)^(p ^ j) * (y.coeff 0)^(p ^ i) :=
+begin
+  calc
+  _ = (verschiebung^[i + j] ((frobenius^[j] x) * (frobenius^[i] y))).coeff (i + j) : _
+... = ((frobenius^[j] x) * (frobenius^[i] y)).coeff 0 : _
+... = (frobenius^[j] x).coeff 0 * ((frobenius^[i] y)).coeff 0 : _
+... = _ : _,
+  { rw iterate_verschiebung_mul },
+  { convert iterate_verschiebung_coeff _ _ _ using 2,
+    rw zero_add },
+  { apply mul_coeff_zero },
+  { simp only [iterate_frobenius_coeff] }
+end
```

Comment:
Is there any reason not to put these in `witt_vector/identities` next to `verschiebung_mul_frobenius` etc? Many of them seem quite general.

Reply:
Fair point. Things are less tangled than when I started the proofs.